WMS: GetFeatureInfo FEATURE_COUNT set connection default

elpaso opened the pull request:
Allow to set a default FEATURE_COUNT (default 10) individually for each WMS connection.

This default also works when adding WMS layers from the browser and it can be overridden for each individual layer when added from the datasource manager.

The deafult user experience when adding layers from the datasource manager is not changed: the default of 10 still applies unless it is explicitly set from the connection settings.

Fix #45206

![immagine](https://github.com/qgis/QGIS/assets/142164/753e21ca-d56a-453b-b4e0-fea151e69ea7)


Funded by: QGIS user group Germany (QGIS Anwendergruppe Deutschland e.V.)

elpaso committed:
WMS: GetFeatureInfo FEATURE_COUNT set connection default

Allow to set a default FEATURE_COUNT (default 10) individually for
each WMS connection.

This default also works when adding WMS layers from the browser and it
can be overridden for each individual layer when added from the
datasource manager.

The deafult user experience when adding layers from the datasource
manager is not changed: the default of 10 still applies unless it
is explicitly set from the connection settings.

Fix #45206

Founded by: QGIS user group Germany (QGIS Anwendergruppe Deutschland e.V.)

elpaso committed:
Spell check

pathmapper reviewed (commented):
Thank you! 

Noticed two things while testing this:

1)
For an **existing** connection, QGIS Browser settings shows the default 10 but there is no FEATURE_COUNT parameter used in the requests.

2)
For a new layer (added via QGIS Browser):

When changing `FEATURE_COUNT` to 33 in Data Source Manager:
<img src="https://github.com/qgis/QGIS/assets/20856381/34c3f657-9485-4c41-a831-ea75af570347" width="400" />

In QGIS Browser it shows `server default`

<img src="https://github.com/qgis/QGIS/assets/20856381/42f3ab1f-18e6-4b3f-9923-dcd2ee112b5e" width="400" />

pathmapper commented on src/ui/qgsnewhttpconnectionbase.ui:
```diff
@@ -140,44 +140,44 @@
+          <widget class="QLabel" name="lblDefaultFeatureCount">
            <property name="text">
-            <string>Ignore GetMap/GetTile/GetLegendGraphic URI reported in capabilities</string>
+            <string>Default maximum number of GetFeatureInfo results</string>
```
```suggestion
            <string>Maximum number of GetFeatureInfo results</string>
```

elpaso replied:
ok "Default" removed, it was to make it clear that this is a connection-level setting that can be overridden when adding the layers from the datasource manager but I guess it is more intuitive this way.

elpaso committed:
Simplify code and set default to 10 for old connections

elpaso commented:
> Thank you!
> 
> Noticed two things while testing this:
> 
> 
> For an **existing** connection, QGIS Browser settings shows the default 10 but there is no FEATURE_COUNT parameter used in the requests.
> 

Yeah, my original intention was to keep the original behavior in case the default was not set explicitly (server default) but then I changed the default to 10, I'll update the logic accordingly.

> 
> For a new layer (added via QGIS Browser):
> 
> When changing `FEATURE_COUNT` to 33 in Data Source Manager: <img alt="" width="400" src="https://private-user-images.githubusercontent.com/20856381/323654409-34c3f657-9485-4c41-a831-ea75af570347.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTA0OTMsIm5iZiI6MTcxMzUxMDE5MywicGF0aCI6Ii8yMDg1NjM4MS8zMjM2NTQ0MDktMzRjM2Y2NTctOTQ4NS00YzQxLWE4MzEtZWE3NWFmNTcwMzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3MDMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlZDI1NGQ3MWQwZTllYTVkYTUwYWUyYTA3MTU0M2M4YzBjNTQ5ODMzN2JlODg2ZTgwYmVkNGEyNGQwOTExNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zOv2FWnZAyYruqKhjIFHKl6KXiwC99IE8MrUVSD2uOc">
> 
> In QGIS Browser it shows `server default`
> <img alt="" width="400" src="https://private-user-images.githubusercontent.com/20856381/323654664-42f3ab1f-18e6-4b3f-9923-dcd2ee112b5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTA0OTMsIm5iZiI6MTcxMzUxMDE5MywicGF0aCI6Ii8yMDg1NjM4MS8zMjM2NTQ2NjQtNDJmM2FiMWYtMThlNi00YjNmLTk5MjMtZGNkMmVlMTEyYjVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3MDMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxN2FkNGUxYTg5YmZlNzZkODNjNWIwMWU1YThlMTgzNzFhNDM5NTkyM2UwYTVkZTNhNzAxODJhYTgzOWY5ZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GUw_gFS5KedUF2tuMgf2AQWb6_9LS1o36gLL_qwRy-8">

I think this is expected: we now have two places where this (max) featureCount is defined and they are supposed to work in a cascading way (from the connection to the individual layer definition), by changing the value in the layer datasource manager select dialog you are not changing the connection level default (this is also why I added "Default" to the label btw) but the modifications only apply to the layers being added.

To change the default value you need to edit the connection settings.

I hope this makes sense, I simplified the code anyway and removed the `default` prefix, 10 is the default in any case if there is no stored setting.

Please have a look to the changes.

elpaso committed:
Update test_qgsowsconnection.py

elpaso committed:
Update test_qgsowsconnection.py

elpaso committed:
Fix double featureCount argument

elpaso commented:
> Thanks for sharing your thoughts and updates.
> 

> What's unexpected (at least for me :-) )

You are right, it should be fixed by https://github.com/qgis/QGIS/pull/57180/commits/c23aaa4c24e7e73f858123725f2d102225490ada

pathmapper commented:
> it should be fixed by https://github.com/qgis/QGIS/commit/c23aaa4c24e7e73f858123725f2d102225490ada

It is, thank you!

DelazJ commented on src/ui/qgsnewhttpconnectionbase.ui:
```diff
@@ -201,10 +208,29 @@
+         <item row="2" column="1">
+          <widget class="QSpinBox" name="sbFeatureCount">
+           <property name="toolTip">
+            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Specify a default value for FEATURE_COUNT when a new layer is created from this connection. &lt;/p&gt;&lt;p&gt;FEATURE_COUNT defines the maximum number of results returned by a GetFeatureInfo request, if not specified the server default value (usually 1) will be used.&lt;/p&gt;&lt;p&gt;Set to 0 to use server default: no FEATURE_COU ...
```
@elpaso if I am correct, with a default value set in that spinbox, this option can't be unspecified. IOW would it be correct to label the tooltip as below?

_Specify a default value for FEATURE_COUNT when a new layer is ~~created~~loaded from this connection.
FEATURE_COUNT defines the maximum number of results returned by a GetFeatureInfo request.~~, if not specified the server default value (usually 1) will be used.~~
Set to 0 to use server default (usually 1): no FEATURE_COUNT parameter will be added to the request._
PS: I didn't check the behavior in QGIS. Comment only based on PR discussion.
